File: openless-all/app/src/lib/ipc/ocr.ts
import type { OcrProvider, DownloadStatus } from "../types"
import { invokeOrMock } from "./shared"
import { mockSettings } from "./mock-data"

/**
 * 后端错误标记：macOS 屏幕录制（TCC）权限未开启。
 * `test_ocr_recognition` 的错误消息以该前缀开头时，设置页应展示系统设置跳转引导。
 */
export const OCR_ERROR_SCREEN_RECORDING_DENIED = "screenRecordingDenied"

/**
 * 后端错误标记：Windows WinRT OCR 缺少中文识别语言包。
 * `test_ocr_recognition` 的错误消息以该前缀开头时，设置页应展示语言包安装引导。
 */
export const OCR_ERROR_WINRT_LANGUAGE_MISSING = "winrtChineseOcrLanguageMissing"

export interface OcrSettings {
    screenContextEnabled: boolean
    activeProviderId: string
    providers: OcrProvider[]
    modelsBaseDir: string
    modelsRootDir: string
    downloadMirror: string
    requiresDownload: boolean
}

export function getOcrSettings(): Promise<OcrSettings> {
    return invokeOrMock(
        "get_ocr_settings",
        undefined,
        () => ({
            screenContextEnabled: mockSettings.screenContextEnabled,
            activeProviderId: mockSettings.activeOcrProvider,
            providers: [
                {
                    id: "disabled",
                    name: "关闭",
                    description: "不注入屏幕上下文",
                    local: true,
                    requiresDownload: false,
                    supportedPlatforms: ["windows", "macos", "linux"],
                },
                {
                    id: "rapidocr",
                    name: "RapidOCR (PP-OCRv6)",
                    description: "本地 ONNX 模型，中文识别质量好，需下载约 30 MB 模型",
                    local: true,
                    requiresDownload: true,
                    supportedPlatforms: ["windows", "macos", "linux"],
                },
            ],
            modelsBaseDir: mockSettings.ocrModelsBaseDir,
            modelsRootDir: "",
            downloadMirror: mockSettings.ocrDownloadMirror,
            requiresDownload: false,
        }),
    )
}

export function setOcrSettings(settings: {
    screenContextEnabled: boolean
    activeProviderId: string
    modelsBaseDir: string
    downloadMirror: string
}): Promise<void> {
    return invokeOrMock(
        "set_ocr_settings",
        {
            screenContextEnabled: settings.screenContextEnabled,
            activeProviderId: settings.activeProviderId,
            modelsBaseDir: settings.modelsBaseDir,
            downloadMirror: settings.downloadMirror,
        },
        () => {
            mockSettings.screenContextEnabled = settings.screenContextEnabled
            mockSettings.activeOcrProvider = settings.activeProviderId
            mockSettings.ocrModelsBaseDir = settings.modelsBaseDir
            mockSettings.ocrDownloadMirror = settings.downloadMirror
        },
    )
}

export function getOcrProviders(): Promise<OcrProvider[]> {
    return invokeOrMock("get_ocr_providers", undefined, () => [])
}

export function startOcrModelDownload(mirror?: string): Promise<void> {
    return invokeOrMock("start_ocr_model_download", { mirror }, () => undefined)
}

export function cancelOcrModelDownload(): Promise<void> {
    return invokeOrMock("cancel_ocr_model_download", undefined, () => undefined)
}

export function getOcrDownloadStatus(): Promise<DownloadStatus[]> {
    return invokeOrMock("get_ocr_download_status", undefined, () => [])
}

export function deleteOcrModels(): Promise<void> {
    return invokeOrMock("delete_ocr_models", undefined, () => undefined)
}

export function testOcrRecognition(): Promise<string | null> {
    return invokeOrMock("test_ocr_recognition", undefined, () => null)
}

File: openless-all/app/src-tauri/src/commands/ocr.rs
//! Tauri commands for OCR / screen-context provider management.

use crate::ocr::{self, DownloadStatus, OcrProvider, rapidocr};
use crate::types::UserPreferences;

use super::CoordinatorState;

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OcrSettings {
    pub screen_context_enabled: bool,
    pub active_provider_id: String,
    pub providers: Vec<OcrProvider>,
    pub models_base_dir: String,
    pub models_root_dir: String,
    pub download_mirror: String,
    pub requires_download: bool,
}

#[tauri::command]
pub fn get_ocr_settings(coord: CoordinatorState<'_>) -> OcrSettings {
    let prefs = coord.prefs().get();
    let providers = ocr::available_providers();
    let requires_download = active_provider_requires_download(&prefs, &providers);
    OcrSettings {
        screen_context_enabled: prefs.screen_context_enabled,
        active_provider_id: prefs.active_ocr_provider.clone(),
        providers,
        models_base_dir: prefs.ocr_models_base_dir.clone(),
        models_root_dir: ocr_models_root(&prefs),
        download_mirror: prefs.ocr_download_mirror.clone(),
        requires_download,
    }
}

#[tauri::command]
pub fn set_ocr_settings(
    coord: CoordinatorState<'_>,
    screen_context_enabled: bool,
    active_provider_id: String,
    models_base_dir: String,
    download_mirror: String,
) -> Result<(), String> {
    let mut prefs = coord.prefs().get();
    prefs.screen_context_enabled = screen_context_enabled;
    prefs.active_ocr_provider = active_provider_id;
    prefs.ocr_models_base_dir = models_base_dir.trim().to_string();
    prefs.ocr_download_mirror = download_mirror.trim().to_string();
    coord.prefs().set(prefs).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_ocr_providers() -> Vec<OcrProvider> {
    ocr::available_providers()
}

#[tauri::command]
pub fn start_ocr_model_download(
    coord: CoordinatorState<'_>,
    mirror: Option<String>,
) -> Result<(), String> {
    let prefs = coord.prefs().get();
    let mirror = mirror.unwrap_or_else(|| prefs.ocr_download_mirror.clone());
    rapidocr::RapidOcrEngine::new(&prefs).start_download(&mirror);
    Ok(())
}

#[tauri::command]
pub fn cancel_ocr_model_download() -> Result<(), String> {
    // 直接操作全局下载状态表，不依赖具体引擎实例或当前偏好。
    rapidocr::cancel_all_downloads();
    Ok(())
}

#[tauri::command]
pub fn get_ocr_download_status() -> Vec<DownloadStatus> {
    rapidocr::all_download_status()
}

#[tauri::command]
pub fn delete_ocr_models(coord: CoordinatorState<'_>) -> Result<(), String> {
    let prefs = coord.prefs().get();
    rapidocr::RapidOcrEngine::new(&prefs)
        .delete_models()
        .map_err(|e| format!("{e:#}"))
}

fn ocr_models_root(prefs: &UserPreferences) -> String {
    // 与 RapidOcrEngine::new 使用同一目录计算，保证设置页展示与实际落盘路径一致。
    rapidocr::RapidOcrEngine::models_dir_for(prefs)
        .display()
        .to_string()
}

fn active_provider_requires_download(prefs: &UserPreferences, providers: &[OcrProvider]) -> bool {
    providers
        .iter()
        .find(|p| p.id == prefs.active_ocr_provider)
        .map(|p| p.requires_download)
        .unwrap_or(false)
}

#[tauri::command]
pub async fn test_ocr_recognition(coord: CoordinatorState<'_>) -> Result<Option<String>, String> {
    let prefs = coord.prefs().get();
    if !prefs.screen_context_enabled {
        return Ok(None);
    }
    ocr::recognize_screen_text(&prefs)
        .await
        .map_err(|e| format!("{e:#}"))
}
